feat(NcCounterBubble): add count and limit props instead of slot

ShGKme opened the pull request:
### ☑️ Resolves

- Fix https://github.com/nextcloud-libraries/nextcloud-vue/issues/5486

Currently, the maximal number is limited via ... ellipsis. It doesn't make sense for numbers. And it is broken on Nextcloud 30 styles.

![image](https://github.com/user-attachments/assets/e02de283-a531-4164-8885-10ecc4ab4173)

In general, having a slot doesn't make much sense for me. As the component is supposed to show a count, it is better to work with it as a number, not any content.

Instead, add an explicit `count` humanize output using:

- Added `count` prop for the value
- Humanize output with localization by default
- Even humanized output from the default slot content, if a plain number is passed
- Mark default slot as deprecated in flavor of `count` or `NcChip`
- Added `raw` prop to disable this behavior
- Improved docs
- Added tests

### 🖼️ Screenshots

#### With new `count` prop

English | Chinese
---|---
![image](https://github.com/user-attachments/assets/d87646c0-c772-4b59-9be3-02ae6f82bd88) | ![image](https://github.com/user-attachments/assets/371cd2f5-e375-4924-ba74-72d43dbaffa5)

#### With old slot

![image](https://github.com/user-attachments/assets/1036ea39-4af1-4495-a4cb-adf7414e2349)

#### Docs for styling

![image](https://github.com/user-attachments/assets/7958adf2-3378-42d8-8050-4a198cb13613)

### 🏁 Checklist

- [x] ⛑️ Tests are included or are not applicable
- [x] 📘 Component documentation has been extended, updated or is not applicable
- [x] 3️⃣ Backport to `next` requested with a Vue 3 upgrade

jancborchardt reviewed (changes requested):
Very nice! :)

Is it possible to have this be default and not a choice/prop? That is, if a counter bubble is used anywhere, it should never go up further than 1000.

> What about using 1K+ instead of 999+?

I'd prefer 999+ as that's more universally localized and a bit clearer that it's "a ton".

ShGKme commented:
> Is it possible to have this be default and not a choice/prop? That is, if a counter bubble is used anywhere, it should never go up further than 1000.

The limit is `999` by default. However, it will work only if you pass the number as a count value, not as a custom HTML content. 

Changing the behavior of the current custom content rendering would be more breaking and also dirty implementation-wise.

susnux commented:
> Is it possible to have this be default and not a choice/prop?

It is 999 by default, which in most cases makes sense, but sometime even 99 would be ok, so we should not hard code this but just have a meaningful default, no?

susnux commented on src/components/NcCounterBubble/NcCounterBubble.vue:
```diff
@@ -8,27 +8,127 @@
 		class="counter-bubble__counter">
-		<slot />
+		<!--
+			@slot The content with the count number.
```
```suggestion
			@slot (deprecated) The content with the count number.
```

Because styleguide output is not that clear I think

ShGKme replied:
It has strikethrough

susnux replied:
> It has strikethrough

I saw this, but it was not clear that this is deprecated or what there is a strike through

susnux commented on src/components/NcCounterBubble/NcCounterBubble.vue:
```diff
@@ -55,6 +155,29 @@ export default {
+		 * Set 0 to disable the limit.
+		 */
+		limit: {
+			type: [Number, String],
```
why allow string? I mean I am fine with that as it easily allows `limit="99"` but I am curious if there is more behind that decision.

ShGKme replied:
To make it easier to pass constants. But I guess, as it is always passed from data, it should be number anyway. So, it can be removed.

ShGKme replied:
Removed

Antreesy commented on src/components/NcCounterBubble/NcCounterBubble.vue:
```diff
@@ -75,9 +205,7 @@ export default {
 	font-size: calc(var(--default-font-size) * .8);
 	overflow: hidden;
 	width: fit-content;
-	max-width: var(--default-clickable-area);
```
I remember we discussed a `min-width` for single-digit values, to make it round

Antreesy reviewed (approved):
Rest is good!

susnux commented on src/components/NcCounterBubble/NcCounterBubble.vue:
```diff
@@ -75,9 +205,7 @@ export default {
 	font-size: calc(var(--default-font-size) * .8);
```
(this is just a note for possible other unrelated TODOs)

ShGKme replied:
Added usage of new variable

marcelklehr commented:
I would like to stress that numbers well above 1000 are a regular occurence. I would like to keep support for strings to allow consumers to pass things like "12.5k" if they so choose

st3iny commented:
As far as I can see, you can still use the slot. It was only deprecated for now. Removing it is a breaking change and implies a major bump.

ShGKme commented:
@marcelklehr do you have an example, where `NcCounterBubble` is used in such a way?

marcelklehr commented:
I use it in the bookmarks app to show bookmarks counts for folders and menu items

jancborchardt commented:
Ok, then let’s do it as you suggested (and also as e.g. Twitter does for favs and retweets):
- "1K" from 1000 on
- "1M" from 1 million on
- Maximum 1 decimal point

Then it fulfills multiple requirements:
- Still keeps an actual count (what @marcelklehr was saying)
- Shows there are a ton of things (what the "999+" was communicating)
- Doesn’t change small digits in large numbers

Sounds good @ShGKme @marcelklehr @susnux?

susnux commented:
Sounds good for me :)

(While I personally would prefer a small `k` but a capital `M` to align with unit prefixes)

jancborchardt commented:
@susnux small "k" and capital "M" sounds good to me too, it seems like the standard. Not sure what the specific reason to Twitter using a capital K is.

ShGKme commented:
@jancborchardt so we remove this feature with limit and add switch to show number in this way instead, correct?

Anyway, I'm going to do it as a new feature with a new prop, not as a slot parser, is that ok? So developers can enable it by switching from slot to component, but the old slot is still available to not break anything.

ShGKme commented:
> @susnux small "k" and capital "M" sounds good to me too, it seems like the standard. Not sure what the specific reason to Twitter using a capital K is.

I'll leave it on localization. It depends on language, and, for example, in Chinese its 万 for 10_000.

en | zh-CN
---|---
![image](https://github.com/user-attachments/assets/e3a38508-2a21-471d-a622-6fe7ee2d0da7) | ![image](https://github.com/user-attachments/assets/0ab3af62-af7d-4822-9b37-cae7cec4b0ab)

jancborchardt commented:
@ShGKme yup, using localization is best. And the example looks great! :100:

ShGKme commented:
> The counts will then still be short due to the magnitude indicators, and no need for settings there.

I'd say, we still need to support old usage. First, some apps already pass `99+` or `1k` which I'd not like to parse these numbers to new format.

Second, some apps have a custom format, for example, Tables in Smart Picker:

![image](https://github.com/user-attachments/assets/38fad576-1626-4799-aae3-2a5f569d1f9c)

jancborchardt commented:
@ShGKme ok, in that case if anything that’s not a number gets passed, it will just be displayed as is? And numbers are subjected to the shortening. Would that work?

ShGKme commented:
> Would that work?

Yes, with my tears =D
(A bit complex implementation)

ShGKme commented:
Before I polish it, @susnux What do you think about this solution?

ShGKme commented:
- Rebased onto `master`
- Changed behavior:
  - Instead of `+` and `limit` -- humanize using Intl compact format
  - Also humanize plain number from slot content

ShGKme commented:
@jancborchardt Updated, see screenshots

susnux commented on src/components/NcCounterBubble/NcCounterBubble.vue:
```diff
@@ -5,34 +5,184 @@
+
+You can use slot to pass any custom content.
 
+Note, if you pass a plain number to the default slot, without raw prop it will be humanized like via `count` prop. This feature is **deprecated** and will be removed in the future. Prefer using `count` prop for numbers with humanization.
```
Shall we deprecate the whole default slot?
As the counter function is done by the prop, using the slot without the counter handling feels like you want to use NcChip or NcUserBubble.

ShGKme replied:
Done
![image](https://github.com/user-attachments/assets/77c8c9dd-421a-4c1e-b9f6-95a3c10399bc)

susnux replied:
BTW just for my reasoning:
From my point of view it makes no sense that we have so many components supporting the same use case.
This one is specialized for *counting* / showing a *count*, so we should stick with that behavior.

If there are complains about this, we still could add a `unit` prop where you can put `n('row', 'rows')` and the result would be `1 row`, `99 rows`, `1k rows`. But lets see if this is needed.

ShGKme commented:
/backport to next

jancborchardt reviewed (approved):
Awesome @ShGKme, sorry for the tears :D

susnux reviewed (approved):
Super cool!

Maybe it makes sense to provide `humanizeCount` as part of `@nextcloud/l10n`?

susnux commented:
(just needs to be rebased)

ShGKme commented:
> Maybe it makes sense to provide `humanizeCount` as part of `@nextcloud/l10n`?

I'm not sure. It's basically a browser feature, just a default compact format + getting locale from `l10n`.

ShGKme commented:
Rebased onto `master` and squashed

ShGKme commented:
> Maybe it makes sense to provide `humanizeCount` as part of `@nextcloud/l10n`?

Maybe makes sense to make it easier for developers
